### DISCUSSION TITLE
How to dao Collection and records

### DISCUSSION THREAD

**Discussion by jeffreymurdock** [pocketbase/pocketbase#988] [Q&A] [answered]
Hello, I'm wondering how the collection records can be used.

Example: `collection, err := app.Dao().FindCollectionByNameOrId("test")`  
this will print:
 `&{{false qjlzy60z5u4tpcm 2022-11-07 20:02:31.671 2022-11-07 20:07:09.632} test false {[0xc0007aac30 0xc0007aac80 0xc000620000 0xc000620050 0xc0006200a0 0xc0006200f0 0xc000620190]} <nil> <nil> <nil> <nil> <nil>} `
 
How to make this with Dao:
```
                type Client struct {
	                url   string 
	                cs    string 
	                ck    string 
                }
                .....
                rows, err := db.Query("SELECT url, cs, ck FROM client")
		if err != nil {
			log.Fatal(err)
		}
		defer rows.Close()
		var temp Client
		for rows.Next() {
			rows.Scan(&temp.url, &temp.cs, &temp.ck)
			credentials := TestClient(temp) /// working with values 
		}
 
```

How to get values from database records to use them in go? 
 
 Tried with no luck: 
```
 type Client struct {
	models.BaseModel
	url   string `db:"url" json:"url"`
	cs    string `db:"cs" json:"cs"`
	ck    string `db:"ck" json:"ck"`
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand the question.

`FindCollectionByNameOrId` will return a collection instance. I guess you want the "test" collection records, right?

You can either get the records as `model.Record` list by using the [record Dao helpers](https://github.com/pocketbase/pocketbase/blob/master/daos/record.go) OR if you want to use your custom model:

1. First make sure that your model satisfies the `model.Model` interface, aka.:
    ```go
    var _ models.Model = (*Client)(nil)

    type Client struct {
        models.BaseModel

        Url string `db:"url" json:"url"`
        Cs  string `db:"cs" json:"cs"`
        Ck  string `db:"ck" json:"ck"`
    }

    func (m *Client) TableName() string {
        return "test" // the name of your collection
    }
    ```
2. To query the database using your custom model, you can have something like:
    ```go
    func FindClient(dao *daos.Dao) (*models.Client, error) {
        model := &Client{}

        err := dao.ModelQuery(model)
            // add some other conditions...
            Limit(1).
            One(model)

        if err != nil {
            return nil, err
        }

        return model, nil
    }
    ```
    _(for other examples you can check the admin dao helpers as a reference - https://github.com/pocketbase/pocketbase/blob/master/daos/admin.go)_

**Reply by ganigeorgiev** [reply]
@jeffreymurdock Oops sorry, my mistake - `models.Client` should be your "yourpackage.Client". The idea is that you are creating an instance of your custom struct and leave the `One()` function to scan its result into it.

**Reply by jeffreymurdock** [reply]
@ganigeorgiev thank you a lot. The last missing puzzle is why the response data from dao to Client struc is empty.
![grafik](https://user-images.githubusercontent.com/25340743/201159317-c806ea7e-140f-482a-ab0f-ae655b0d15c6.png)
```
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		re, err := FindClient(app.Dao())
		if err != nil {
			return err
		}
		log.Println(re.BaseModel.Id)
		log.Println(re.url)
		log.Println(re.cs)
		log.Println(re.ck)
		return nil
	})
```
Prints: 
```
2022/11/10 16:58:53 ymveh6yw53kyd8r
2022/11/10 16:58:53
2022/11/10 16:58:53
2022/11/10 16:58:53
```
Is there something missing ?

**Reply by ganigeorgiev** [reply]
Sorry I didn't test the example I just copied your initial Client struct. The struct fields must be "public/exported" aka.:

```go
type Client struct {
    models.BaseModel

    Url string `db:"url" json:"url"`
    Cs  string `db:"cs" json:"cs"`
    Ck  string `db:"ck" json:"ck"`
}
```
